File: typescript/Y2022/Day1.ts
import { Solution } from '../run';

const part1 = (input: string) => {
  const rows: string[] = input.trim().split('\n\n');

  return Math.max(
    ...rows.map((row) => row.split('\n').reduce((acc, val) => acc + +val, 0)),
  );
};

const part2 = (input: string) => {
  const rows: string[] = input.trim().split('\n\n');

  const [a, b, c] = rows
    .map((row) => row.split('\n').reduce((acc, val) => acc + +val, 0))
    .sort((a, b) => b - a);

  return a + b + c;
};

const solution: Solution<number, number> = {
  part1,
  part2,
  answers: [65912, 195625],
};

export default solution;

File: src/y2022/day_1.rs
use crate::Solution;

pub fn get_solution() -> Solution<usize, usize> {
    Solution {
        part_1,
        part_2,
        answer: (65_912, 195_625),
    }
}

fn part_1(input: &str) -> anyhow::Result<usize> {
    let lines = input.lines();

    let mut max = 0;
    let mut current = 0;
    for line in lines {
        if line.is_empty() {
            max = max.max(current);
            current = 0;
        } else {
            let i = line.parse::<usize>().unwrap();
            current += i;
        }
    }

    Ok(max)
}

fn part_2(input: &str) -> anyhow::Result<usize> {
    let lines = input.lines();

    let mut max = vec![];
    let mut current = 0;
    for line in lines {
        if line.is_empty() {
            max.push(current);
            current = 0;
        } else {
            let i = line.parse::<usize>().unwrap();
            current += i;
        }
    }

    max.sort_unstable_by(|a, b| b.partial_cmp(a).unwrap());

    Ok(max[0] + max[1] + max[2])
}
